cross_validation_mpi.py

```diff
@@ -0,0 +1,173 @@
+from sklearn.model_selection import LeaveOneGroupOut
+from sklearn.metrics import confusion_matrix
+from sklearn.metrics import accuracy_score, f1_score
+from sklearn.ensemble import RandomForestClassifier as RandomForestClassifier
+from sklearn.decomposition import PCA
+from sklearn.pipeline import Pipeline
+#from sklearn.neural_network import MLPClassifier as MLPClassifier
+#from sklearn.ensemble import AdaBoostClassifier as AdaBoostClassifier
+#from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis as QuadraticDiscriminantAnalysis
+
+
+def naive_gen():
+    class NaiveModel:
+        def fit(self,X_train,y_train):
+            self.majority = y_train.value_counts().idxmax()
+        def predict(self,X_test):
+            y_hat = [self.majority for i in range(len(X_test))]
+            return y_hat
+    model = NaiveModel()
+    return model
+    
+def rf_pca(random_state_1=15,random_state_2=20):
+    #defines the Random Foest PCA model
+    model_name = 'RF_PCA_lower_border'
+    pca = PCA(n_components='mle',random_state=random_state_1)
+    rf = RandomForestClassifier(random_state=random_state_2)
+    model = Pipeline(steps=[('pca',pca),('rf',rf)])  
+    return model
+  
+def save_results(results, conf_matrix, model_name: str):
+    #save the results to a csv file
+    results.to_csv(model_name+'_final_cross_val.csv',index=False)
+    filename = model_name+'_matrix_mean.pkl'
+    fileObject = open(filename, 'wb')
+    pickle.dump(conf_matrix, fileObject)
+    fileObject.close()
+    
+def balance_data(X_train, y_train, sampling_method, random_state=23):
+    begin = time.time()
+    k_neighbors = y_train.value_counts().min()-1
+    if (k_neighbors <= 1):
+        raise ValueError("Not enough data to rebalance. Must be more than 1.")
+    if (sampling_method == 'BorderlineSMOTE'):
+        smote = BorderlineSMOTE(sampling_strategy='not majority',k_neighbors=k_neighbors,random_state=random_state)
+    elif (sampling_method == 'ADASYN'):
+        smote = ADASYN(sampling_strategy='not majority',n_neighbors=k_neighbors,random_state=random_state)
+    elif (sampling_method == 'SMOTE'):
+        smote = SMOTE(sampling_strategy='not majority',k_neighbors=k_neighbors,random_state=random_state)
+    else:
+        raise ValueError("Sampling method not available")
+    #rebalance the data
+    X_train, y_train = smote.fit_resample(X_train, y_train)
+    end = time.time()
+    print("Time to rebalance: "+str(np.round(end-begin,3)))
+    return X_train, y_train
+     
+def fit_predict_model(X_data, y, train_ind, test_ind, model, balance=True, rank=None):
+    #now fit using the indeces given by the kfold splitter
+    X_train = X_data[train_ind]
+    y_train = y.iloc[train_ind]
+    #get the labels for the confusion matrix
+    if (balance == True):
+        X_train, y_train = balance_data(X_train, y_train, 'SMOTE')     
+    #fit on  data
+    model.fit(X_train,y_train)
+    del X_train
+    del y_train
```

Comment:
this will happen automatically with python garbage collection when this function returns.